### PR TITLE
Add busy indicator to mini reporter

### LINKED ISSUE #598: Default reporter should show a busy indicator
Some people have been confused and thought AVA froze, just because it was stuck on showing one test title for a while. This can happen when a lot of async tests are run concurrently and it takes some time for them all to finish.

I'm thinking something simple like:

untitled12

Or

untitled10

From: https://github.com/sindresorhus/cli-spinners/blob/master/spinners.json

### PR DESCRIPTION
**Pull request by jamestalmage**
Add a spinner to the mini reporter to indicate ongoing test progress.

Fixes #598.

### CODE REVIEW

**Comment by jamestalmage** [comment]
ava-spinner-demo

I ran all the visual tests and they all look good on iTerm2, OSX El Capitan.

**Comment by sotojuan** [comment]
This is awesome, thanks!

File: lib/reporters/mini.js [hunk 1]
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
...
+	this.spinnerFrame = 0;
 	this.lastLineTracker = lastLineTracker();
 };
 
+MiniReporter.prototype.spinnerChar = function () {
+	return this.spinnerFrames[this.spinnerFrame];
+};
+
+MiniReporter.prototype.clearInterval = function () {
+	if (this.interval) {
```
Comment (sindresorhus) [inline lib/reporters/mini.js]: Is there really any point in guarding this? `clearInterval` gracefully accepts anything.

File: lib/reporters/mini.js [hunk 2]
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
 module.exports = MiniReporter;
 
 MiniReporter.prototype.start = function () {
-	return '';
+	var self = this;
+	this.interval = setInterval(function () {
+		self.spinnerFrame = (self.spinnerFrame + 1) % self.spinnerFrames.length;
+		self.write(self.spinnerChar() + self.currentTest);
```
Comment (sindresorhus) [inline lib/reporters/mini.js]: Can you wrap the spinner char in `chalk.dim()`?

Comment (sindresorhus) [inline lib/reporters/mini.js]: It's also a bit weird that `self.spinnerChar() + self.currentTest` is duplicated in `MiniReporter#test` too.

**Comment by sindresorhus** [comment]
Now the test title and the total count (71 passed) is vertically unaligned. They should be aligned.

File: lib/reporters/mini.js [hunk 3]
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
...
+	this.clearInterval();
 	this.passCount = 0;
 	this.failCount = 0;
 	this.skipCount = 0;
 	this.rejectionCount = 0;
 	this.exceptionCount = 0;
 	this.currentStatus = '';
+	this.currentTest = '';
 	this.statusLineCount = 0;
+	this.spinnerFrame = 0;
```
Comment (sindresorhus) [inline lib/reporters/mini.js]: `spinnerIndex` would be a better name. When it says `spinnerFrame` I thought it was the actual frame.

**Comment by sindresorhus** [comment]
While you're at it, can you put a linebreak over the test title to give it some breathing-room?

File: lib/reporters/mini.js [hunk 4]
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
...
 
+MiniReporter.prototype.spinnerChar = function () {
+	return this.spinnerFrames[this.spinnerFrame];
+};
+
+MiniReporter.prototype.clearInterval = function () {
+	if (this.interval) {
+		clearInterval(this.interval);
+	}
+	this.interval = 0;
```
Comment (novemberborn) [inline lib/reporters/mini.js]: Reset to `null` since Node uses object refs?

File: lib/reporters/mini.js [hunk 5]
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
...
+};
+
 MiniReporter.prototype.test = function (test) {
-	var status = '';
+	this.currentTest = this._test(test);
```
Comment (novemberborn) [inline lib/reporters/mini.js]: Should this be `currentStatus` instead?

File: lib/reporters/mini.js [hunk 6]
```diff
@@ -10,6 +11,10 @@ function MiniReporter() {
 		return new MiniReporter();
 	}
 
+	var spinnerDef = spinners.dots;
+	this.spinnerFrames = spinnerDef.frames;
+	this.spinnerInterval = spinnerDef.interval;
```
Comment (novemberborn) [inline lib/reporters/mini.js]: How about using `var dots = require('cli-spinners').dots` above and just referencing `dots.frames`/`dots.interval`?

File: lib/reporters/mini.js (hunk 2 above)
Comment (novemberborn) [inline lib/reporters/mini.js]: Could abstract that into `writeStatusWithSpinner()` or something.

Comment (jamestalmage) [inline lib/reporters/mini.js]: > Can you wrap the spinner char in chalk.dim()?

I just gave it a try, I can't tell the difference.

File: lib/reporters/mini.js (hunk 5 above)
Comment (jamestalmage) [inline lib/reporters/mini.js]: `currentStatus` includes the spinner character. currentTest does not.

**Comment by jamestalmage** [comment]
Should pass / fail count be on the same line?

A: 

screenshot 2016-03-05 21 40 58

vs.

B:

screenshot 2016-03-05 21 40 00

**Comment by jamestalmage** [comment]
OK,
I incorporated most of the PR feedback.

Also, I switched it to option A as described above (counts on separate lines). That is the last commit, so it can be dropped if there is not agreement.

File: lib/reporters/mini.js (hunk 6 above)
Comment (jamestalmage) [inline lib/reporters/mini.js]: @sindresorhus usually objects to that. I don't have a strong opinion either way.

Comment (jamestalmage) [inline lib/reporters/mini.js]: Also, this way leaves us open to having a command line option to configure your spinner:

```
$ ava --spinner=balloon
```

then

```js
var spinnerDef = spinners[config.spinner];
```

File: lib/reporters/mini.js (hunk 2 above)
Comment (sindresorhus) [inline lib/reporters/mini.js]: @jamestalmage It's not supported on all Terminals, but it dims the spinner to be less intruding. Could maybe also do `chalk.gray.dim()` to support all terminals.

**Comment by sindresorhus** [comment]
One more minor tweak. Move the spinner one char to the right without moving anything else. It's a bit too far from the test title and a bit too much crammed into the left edge.

`⠋  test title` => ` ⠋ test title`

**Comment by sindresorhus** [comment]
I'm good with `A`, but will let @vdemedes chime in.

File: lib/reporters/mini.js [hunk 7]
```diff
@@ -51,15 +85,18 @@ MiniReporter.prototype.test = function (test) {
...
 	if (this.passCount > 0) {
-		status += '  ' + colors.pass(this.passCount, 'passed');
+		status += '   ' + colors.pass(this.passCount, 'passed');
 	}
 
 	if (this.failCount > 0) {
-		status += '  ' + colors.error(this.failCount, 'failed');
+		if (this.passCount > 0) {
+			status += '\n';
+		}
```
Comment (sindresorhus) [inline lib/reporters/mini.js]: Wouldn't you need to do this for `skipped` and `todo` tests results too?

**Comment by vdemedes** [comment]
I'm also ok with A (assuming all other counters will be updated too).

File: lib/reporters/mini.js (hunk 7 above)
Comment (jamestalmage) [inline lib/reporters/mini.js]: Good call on `todo`. I implemented this before rebasing, so yeah - that almost certainly needs a fix.

Are skipped tests shown yet? I thought there was still a pending PR for that.

Comment (jamestalmage) [inline lib/reporters/mini.js]: Oh - and I also missed everything in `final`

Comment (sindresorhus) [inline lib/reporters/mini.js]: Yes, with master:

screen shot 2016-03-06 at 14 49 01

**Comment by jamestalmage** [comment]
All changes made.

File: lib/reporters/mini.js [hunk 8]
```diff
@@ -18,10 +26,16 @@ function MiniReporter() {
 module.exports = MiniReporter;
 
 MiniReporter.prototype.start = function () {
-	return '';
+	var self = this;
+	this.interval = setInterval(function () {
+		self.spinnerIndex = (self.spinnerIndex + 1) % self.spinnerFrames.length;
+		self.write(self.prefix());
+	}, this.spinnerInterval);
+	return this.prefix('');
```
Comment (novemberborn) [inline lib/reporters/mini.js]: The `''` seems unnecessary here.

File: lib/reporters/mini.js (hunk 5 above)
Comment (novemberborn) [inline lib/reporters/mini.js]: Ah didn't realize there was already a `currentStatus` property.

**Comment by novemberborn** [comment]
Cool!

**Comment by sindresorhus** [comment]
The default reporter looks so beautiful now 😍. Thanks James :)

**Pull request closed without merging**